controller/src/main/java/io/pravega/controller/store/stream/tables/serializers/StateRecordSerializer.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.stream.tables.serializers;
+
+import io.pravega.common.io.serialization.RevisionDataInput;
+import io.pravega.common.io.serialization.RevisionDataOutput;
+import io.pravega.common.io.serialization.VersionedSerializer;
+import io.pravega.controller.store.stream.tables.State;
+import io.pravega.controller.store.stream.tables.StateRecord;
+
+import java.io.IOException;
+
+public class StateRecordSerializer extends VersionedSerializer.WithBuilder<StateRecord, StateRecord.StateRecordBuilder> {
+    @Override
+    protected byte getWriteVersion() {
+        return 0;
+    }
+
+    @Override
+    protected void declareVersions() {
+        version(0).revision(0, this::write00, this::read00);
+    }
+
+    private void read00(RevisionDataInput revisionDataInput, StateRecord.StateRecordBuilder builder) throws IOException {
+        int ordinal = revisionDataInput.readCompactInt();
+        if (ordinal < State.values().length) {
+            builder.state(State.values()[ordinal]);
+        } else {
+            // TODO: what happens if newer states are added and we are trying to deserialize a higher revisioned data from
```

Comment:
Then, when you add newer values to the enum, you need to make sure you encode those newer values in a newer revision and never in the older revision. The old code will not read the newer values, but newer code will simply use the value from the higher revision (yes, you are double-encoding here).

You'll need to figure out how to handle that scenario in older code though, since you may not have an equivalent value to map to.

Reply:
The concern is around how to handle the scenario. A new State implies some ongoing processing. And we cannot change the state to UNKNOWN or some default if read from older code while performing some processing. That can render stream metadata inconsistent.